demos/http/http_demo_s3_download_multithreaded/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,1022 @@
+    memcpy( serverHost, pHost, hostLen );
+
+    /* Initialize server information. */
+    serverInfo.pHostName = serverHost;
+    serverInfo.hostNameLength = hostLen;
+    serverInfo.port = HTTPS_PORT;
+
+    /* Establish a TLS session with the HTTP server. This example connects
+     * to the HTTP server as specified in SERVER_HOST and HTTPS_PORT
+     * in demo_config.h. */
+    LogInfo( ( "Establishing a TLS session with %s:%d.",
+               serverHost,
+               HTTPS_PORT ) );
+
+    opensslStatus = Openssl_Connect( pNetworkContext,
+                                     &serverInfo,
+                                     &opensslCredentials,
+                                     TRANSPORT_SEND_RECV_TIMEOUT_MS,
+                                     TRANSPORT_SEND_RECV_TIMEOUT_MS );
+
+    return opensslStatus == OPENSSL_SUCCESS ? EXIT_SUCCESS : EXIT_FAILURE;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool downloadS3ObjectFile( const char * pHost,
+                                  const size_t hostLen,
+                                  const char * pRequest,
+                                  const size_t requestLen,
+                                  mqd_t requestQueue,
+                                  mqd_t responseQueue )
+{
+    bool returnStatus = true;
+    QueueOpStatus_t queueOpStatus = QUEUE_OP_SUCCESS;
+
+    size_t requestCount = 0;
+
+    /* Configurations of the initial request headers. */
+    HTTPRequestInfo_t requestInfo = { 0 };
+
+    /* Response data sent over queue. */
+    ResponseItem_t responseItem = { 0 };
+
+    /* The length of the file at S3_PRESIGNED_GET_URL. */
+    size_t fileSize = 0;
+
+    /* The number of bytes we want to request within each range of the file. */
+    size_t numReqBytes = 0;
+    /* curByte indicates which starting byte we want to download next. */
+    size_t curByte = 0;
+
+    /* Initialize the request object. */
+    requestInfo.pHost = pHost;
+    requestInfo.hostLen = hostLen;
+    requestInfo.method = HTTP_METHOD_GET;
+    requestInfo.methodLen = HTTP_METHOD_GET_LENGTH;
+    requestInfo.pPath = pRequest;
+    requestInfo.pathLen = requestLen;
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. This is done in
+     * order to download the file in parts. */
+    requestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Get the length of the S3 file. */
+    returnStatus = getS3ObjectFileSize( &requestInfo,
+                                        requestQueue,
+                                        responseQueue,
+                                        &fileSize );
+
+    if( returnStatus == true )
+    {
+        /* Ensure the requested amount will fit in the buffer. */
+        if( fileSize < USER_BUFFER_LENGTH - HTTP_MAX_RESPONSE_HEADERS_SIZE_BYTES )
```

Comment:
If use USER_BUFFER_LENGTH configured is less than HTTP_MAX_RESPONSE_HEADERS_SIZE_BYTES, then you will get overflow here.
Also HTTP_MAX_RESPONSE_HEADERS_SIZE_BYTES is not really intended to be used this way. It is an overarching max number of headers accepted, not necessarily what we expect from S3.